lib/mix/lib/mix/dep/loader.ex

```diff
@@ -320,6 +322,13 @@ defmodule Mix.Dep.Loader do
     end
   end
 
+  defp incompatible_requirement?(%Mix.Dep{opts: opts}) do
+    case opts[:lock] do
+      {scm, _, _, _} -> is_nil(opts[scm])
```

Comment:
> Would making this work for the other cases be good enough

It isn't good enough because any developer can add a new SCM any time. We should never look into `opts[:lock]`, only the SCM itself is allowed to do that.